### PR TITLE
[testing framework] allow tests to define and override providers

### PR DESCRIPTION
**Pull request by liamcervante**
This PR gives test files and run blocks the ability to set and/or override the provider settings in modules under test.

The approach taken here is to introduce a `TransformForTest` function into the `configs` package that will copy provider blocks from the test file into the config under test as if they had been written there. Then when Terraform executes it simply reads the provider blocks from the configuration.

There is an update to the init function as well, to make sure that when it is gathering provider requirements and downloads it also looks into testing modules that are referenced from test files to see if there are any required provider blocks in these. It doesn't actually need to look into the test files at all, since they don't contain provider requirements or definitions themselves they only allow providers within actual config to be overridden. If someone tries to define a provider in a test file that isn't in any actual config file, the soon-to-be-updated validate command will catch that and prevent the test from executing anyway.

An alternate approach here would be to update the plan options to include an external `ProviderOverrides` field. Previously for the testing functionality I have avoided modifying the graph operation as much as possible, and only adding in external inputs where there is no other option. In this case, I do have the option of editing the config directly first so that is the approach taken. Having said that, feedback here saying this that updating the plan options is actually a better approach is fine, so let me know if that would be better. In my opinion though I want to avoid tests and actual plan/apply operations being too different.

---

This PR is part of a chain of PRs adding support for the *beta* functionality of the testing framework.

|            Description            | Branch                       | PR     |
|:---------------------------------:|------------------------------|--------|
| Preperation / common refactoring  | liamcervante/tests/prep      | #33445 |
| Implement `module` block          | liamcervante/tests/module    | #33456 |
| Implement `providers` block       | liamcervante/tests/providers | #33466 |
| Implement interrupt functionality | liamcervante/tests/cancel    |   #33477 |

The above PRs depend on one another, and should be reviewed and merged in order. If you want to see all the changes at once you can navigate to the last PR and update the branch it is to be merged into to `main` and this will show the summary of changes.

### CODE REVIEW

File: internal/configs/config.go [hunk 1]
```diff
@@ -487,45 +522,75 @@ func (c *Config) addProviderRequirements(reqs getproviders.Requirements, recurse
...
-					Detail:  fmt.Sprintf("Incorrect version constraint syntax: %s.", err.Error()),
-					Subject: provider.Version.DeclRange.Ptr(),
-				})
+		moreDiags := c.addProviderRequirementsFromProviderBlock(reqs, provider)
+		diags = append(diags, moreDiags...)
+	}
+
+	// We may have provider blocks and required_providers set in some testing
+	// files.
+	if tests {
```
Comment (jbardin) [inline internal/configs/config.go:531]: I'm not quite sure I understand this, shouldn't the requirements be the same regardless of whether there are providers configured within tests?

Comment (jbardin) [reply internal/configs/config.go:531]: I don't think you can have different provider requirements. We need to have a stable set defined in the provider lock file, plus you wouldn't want the tests running with a different provider version invalidating the test results. There could be _additional_ provider defines in the test block, but those need to be resolved already. (from the Go perspective, `go.mod` must always accounts for test dependencies too, and you wouldn't run `go test` with a different `go.mod`)

Comment (liamcervante) [reply internal/configs/config.go:531]: Yeah, all of this is being read and considered during `terraform init`. 

So any provider requirements set in the test dependencies (whether in dedicated test modules or provider blocks) will be checked, and additional providers downloaded, when the user runs `terraform init`. It's all just extra requirements that the provider fetcher needs to take into account when getting the providers and building the lock file.

There won't be different requirements, it's just additional requirements that need to be accounted for. If the tests say I need a provider of version x and the configuration says I need a provider of version y then the init command will fail and say I can't find a provider with a valid version, as it does at the moment if you included two modules in your configuration with incompatible requirements.

Hopefully that makes sense, happy to have a quick call to discuss if needed!

Comment (liamcervante) [reply internal/configs/config.go:531]: In general, users shouldn't be setting provider versions directly in the test files (as they shouldn't be in real configuration as well) as that attribute is deprecated. 

But they can write modules that are only loaded by the tests that might pull in providers that aren't needed by the main configuration, so those modules need to be processed by `terraform init` for provider requirements, but if those requirements are incompatible with the config under test then that's a problem for the user to fix, and init will tell them they have incompatible version requirements.

Comment (jbardin) [reply internal/configs/config.go:531]: Thanks! I think I had a false assumption about the ordering of operations in my head last week. As long as all provider requirements are always accounted for this looks good.

**Pull request merged**